sdk/core/Azure.Core/src/Shared/TaskExtensions.cs

```diff
@@ -133,5 +157,69 @@ private static void VerifyTaskCompleted(bool isCompleted)
             public void Dispose() => _asyncEnumerator.DisposeAsync().EnsureCompleted();
 #pragma warning restore AZC0107 // Do not call public asynchronous method in synchronous scope.
         }
+
+        public readonly struct WithCancellationAwaitable<T>
+        {
+            private readonly CancellationToken _cancellationToken;
+            private readonly ConfiguredTaskAwaitable<T> _awaitable;
+
+            public WithCancellationAwaitable(Task<T> task, CancellationToken cancellationToken)
+            {
+                _awaitable = task.ConfigureAwait(false);
+                _cancellationToken = cancellationToken;
+            }
+
+            public WithCancellationAwaiter<T> GetAwaiter() => new WithCancellationAwaiter<T>(_awaitable.GetAwaiter(), _cancellationToken);
+        }
+
+        public readonly struct WithCancellationAwaiter<T> : ICriticalNotifyCompletion
+        {
+            private readonly CancellationToken _cancellationToken;
+            private readonly ConfiguredTaskAwaitable<T>.ConfiguredTaskAwaiter _taskAwaiter;
+
+            public WithCancellationAwaiter(ConfiguredTaskAwaitable<T>.ConfiguredTaskAwaiter awaiter, CancellationToken cancellationToken)
+            {
+                _taskAwaiter = awaiter;
+                _cancellationToken = cancellationToken;
+            }
+
+            public bool IsCompleted => _taskAwaiter.IsCompleted || _cancellationToken.IsCancellationRequested;
+
+            public void OnCompleted(Action continuation)
+                => _taskAwaiter.OnCompleted(CreateContinuation(continuation));
+
+            public void UnsafeOnCompleted(Action continuation)
+                => _taskAwaiter.UnsafeOnCompleted(CreateContinuation(continuation));
+
+            public T GetResult()
+            {
+                Debug.Assert(IsCompleted);
+                if (!_taskAwaiter.IsCompleted)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+                }
+                return _taskAwaiter.GetResult();
+            }
+
+            private Action CreateContinuation(in Action originalContinuation)
+            {
+                if (!_cancellationToken.CanBeCanceled || _taskAwaiter.IsCompleted)
+                {
+                    return originalContinuation;
+                }
+
+                CancellationTokenRegistration registration = _cancellationToken.Register(originalContinuation);
+                Action reference = originalContinuation;
+                return () =>
+                {
+                    Action continuation = Interlocked.Exchange(ref reference, null);
```

Comment:
Why do we need `Interlocked.Exchange` here? To prevent double registration?

Reply:
I see, it's so we don't run continuation twice.

Reply:
I'm not sure it works though.

In cancellation scenario both `originalContinuation` would be invoked first and if Task completes too the lambda would run next invoking `originalContinuation` again.

Reply:
Here is the code generated for [this closure](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQBUAWATgKYCGwAlgHYDmA3MmpjgEoCuVALhQLbHYBhAPY8ADhQA2xQgGVpANwoBjYgGcGSRgGZ0qzoTZLO6AOoVO+AaSoqJE0tyFUAggHdS50gCMpAHlwAfOgA3sjo4eiihBTyDsToJOROEgCe6FY2xHYOFE64QgDWxFToAPpK1rb2jlT5RVQaEZHRsZzxicDJacJUAGYUNGwkwFAAbG4enN5+gYJO/YPDYxPm0mVTqgUrbYQaYRFRMXHMMOmJbT3cVGw5TgAU1CfoQtE01KQSl9Q3NQCU++FQkgmk0KL10HcAITlSpZaq5WqFYqCawAIWIGVsxGA6AgEHWpE222k2AAkqphGIpG1gP9gSDAQCGREoAB2Z6vd6fJxXH4IxrMgC+yCZTUxcNuiPqLGIbz0hElCVlFHlioAvGUKplsjU6siZXKdncXgMuV9rpLfgKGVhTiRetJiip0BqTW8qB9zXynNaQWyIb8XQFRREgcy/XBTkoed91ehSVxpBIhEoisBsABRAAeSnw1hoxDu9qVDpImQANOhrnYrSGGWCIdGuLGauhIRrqxI6eGmmGewySIaFTVsAARFWiISqQu1+n9ppN3mSu6z+foYVzkGC30bwVAA=):
```cs
private Action CreateContinuation([In] [IsReadOnly] ref Action originalContinuation)
{
    <>c__DisplayClass2_0 <>c__DisplayClass2_ = new <>c__DisplayClass2_0();
    if (!_cancellationToken.CanBeCanceled)
    {
        return originalContinuation;
    }
    <>c__DisplayClass2_.registration = _cancellationToken.Register(originalContinuation);
    <>c__DisplayClass2_.reference = originalContinuation;
    return new Action(<>c__DisplayClass2_.<CreateContinuation>b__0);
}
```
```cs
private sealed class <>c__DisplayClass2_0
{
    public Action reference;
    public CancellationTokenRegistration registration;
    internal void <CreateContinuation>b__0()
    {
        Action action = Interlocked.Exchange(ref reference, null);
        if (action != null)
        {
            registration.Dispose();
            action();
         }
    }
}
```
`originalContinuation` is assigned to the `reference`, and then `reference` is replaced with `null`, so it is called only once. `registration.Dispose()` can be called twice but that is ok.

Reply:
> originalContinuation is assigned to the reference, and then reference is replaced with null

Yes, but you are also using the `originalContinuation` as the cancellation token callback which would fire unconditionally (it's never replaced by null). Cancellation token callback doesn't clear the `reference` so continuation would still run the `originalContinuation` once.

What am I missing?

Reply:
No, you're actually right. That's a great catch. Fixed and added tests.